lib/internal/modules/esm/hooks.js

```diff
@@ -143,42 +155,74 @@ class Hooks {
    * @param {string} urlOrSpecifier
    * @param {string} parentURL
    * @param {any} [data] Arbitrary data to be passed from the custom
+   * @param {number} [chainId]
+   * @param {number} [threadId]
+   * @param {any} [workerData]
    * loader (user-land) to the worker.
    */
-  async register(urlOrSpecifier, parentURL, data) {
+  async register(urlOrSpecifier, parentURL, data, chainId, threadId, workerData) {
     const cascadedLoader = require('internal/modules/esm/loader').getOrInitializeCascadedLoader();
     const keyedExports = await cascadedLoader.import(
       urlOrSpecifier,
       parentURL,
       kEmptyObject,
     );
-    await this.addCustomLoader(urlOrSpecifier, keyedExports, data);
+    await this.addCustomLoader(urlOrSpecifier, keyedExports, data, chainId, threadId, workerData);
   }
 
+
   /**
    * Collect custom/user-defined module loader hook(s).
    * @param {string} url Custom loader specifier
    * @param {Record<string, unknown>} exports
-   * @param {any} [data] Arbitrary data to be passed from the custom loader (user-land)
-   * to the worker.
+   * @param {any} [data] Arbitrary data to be passed from the custom loader (user-land) to the worker.
+   * @param {number} [chainId]
+   * @param {number} [threadId]
+   * @param {any} [workerData]
    * @returns {any | Promise<any>} User data, ignored unless it's a promise, in which case it will be awaited.
    */
-  addCustomLoader(url, exports, data) {
+  addCustomLoader(url, exports, data, chainId, threadId, workerData) {
     const {
       initialize,
       resolve,
       load,
     } = pluckHooks(exports);
 
+    if (chainId) {
+      this.#chains[chainId] ??= {
+        resolve: [...defaultChains.resolve],
+        load: [...defaultChains.load],
+        registered: [],
+      };
+    }
+
+    const chain = this.getChain(chainId);
+
+    // Do not register the same hook with the same data
+    for (const hook of chain.registered) {
+      if (hook.url === url && (typeof data === 'undefined' || isDeepStrictEqual(hook.data, data))) {
+        return undefined;
+      }
+    }
+
+    /**
+     * At the moment the chains are new cleared, especially since there is no de-register API.
+     * In the future we might want to implement reference-counting so that when all threads
+     * using a specific chain are terminated we can eventually dispose unused chains.
+     */
+    ArrayPrototypePush(chain.registered, { url, data });
+
     if (resolve) {
-      const next = this.#chains.resolve[this.#chains.resolve.length - 1];
-      ArrayPrototypePush(this.#chains.resolve, { __proto__: null, fn: resolve, url, next });
+      const next = chain.resolve[chain.resolve.length - 1];
```

Comment:
Should we use `ArrayPrototypeAt()` with an argument of `-1`?

Reply:
I don't have a strong opinion. This one might be faster.